keepers/keeper-core/src/lib.rs

```diff
@@ -298,107 +248,135 @@ async fn parallel_confirm_transactions(
+) -> Result<Vec<Result<(), SendTransactionError>>, TransactionExecutionError> {
+    let mut results = vec![Err(SendTransactionError::ExceededRetries); transactions.len()];
+    let mut retries = 0;
+
+    if transactions.is_empty() {
+        return Ok(results);
     }
 
-    let instructions_per_tx = calculate_instructions_per_tx(client, &instructions, signer)
+    let mut signed_txs = sign_txs(client, transactions, signer)
         .await
-        .map_err(|e| {
-            TransactionExecutionError::ClientError(e.to_string(), instructions.to_vec())
-        })?;
+        .map_err(|e| TransactionExecutionError::ClientError(e.to_string()))?;
 
-    for _ in 0..retry_count {
-        let mut executed_signatures: HashMap<Signature, usize> = HashMap::new();
-        let instruction_batches: Vec<&[Instruction]> =
-            instructions.chunks(instructions_per_tx).collect();
-        parallel_submit_transactions(
-            client,
-            signer,
-            &instruction_batches,
-            &mut executed_signatures,
-        )
-        .await?;
+    while retries < retry_count {
+        let mut submitted_signatures = HashMap::new();
+
+        for (idx, tx) in signed_txs.iter().enumerate() {
+            if results[idx].is_ok() {
+                continue;
+            }
+
+            // Future optimization: submit these in parallel batches and refresh blockhash for every batch
+            match client.send_transaction(tx).await {
+                Ok(signature) => {
+                    submitted_signatures.insert(signature, idx);
+                }
+                Err(e) => match e.get_transaction_error() {
+                    Some(TransactionError::BlockhashNotFound)
+                    | Some(TransactionError::AlreadyProcessed) => {
+                        submitted_signatures.insert(tx.signatures[0], idx);
+                    }
+                    Some(_) | None => {
+                        warn!("Transaction error: {}", e.to_string());
+                        results[idx] = Err(SendTransactionError::TransactionError(e.to_string()))
+                    }
+                },
+            }
+        }
 
         tokio::time::sleep(Duration::from_secs(confirmation_time)).await;
 
-        parallel_confirm_transactions(client, &mut executed_signatures).await;
+        for idx in parallel_confirm_transactions(client, submitted_signatures)
+            .await
+            .into_values()
+        {
+            results[idx] = Ok(());
+        }
 
-        // All have been executed
-        if executed_signatures.is_empty() {
-            return Ok(());
+        if results.iter().all(|r| r.is_ok()) {
+            break;
         }
 
-        // Update instructions to the ones remaining
-        instructions = executed_signatures
-            .into_values()
-            .flat_map(|i| instruction_batches[i])
-            .cloned()
-            .collect::<Vec<_>>();
+        // Re-sign transactions with fresh blockhash
```

Comment:
going to send duplicates if always resigning. should just need to resign once the blockhash you previously signed txs with expires based on processed commit level. there's a is_blockhash_valid_with_config method

Reply:
added logic to do less re-signing